src/StoreApi/Utilities/CartController.php

```diff
@@ -213,17 +215,144 @@ public function validate_add_to_cart( \WC_Product $product, $request ) {
 		do_action( 'wooocommerce_store_api_validate_add_to_cart', $product, $request );
 	}
 
+	// phpcs:disable Squiz.Commenting.FunctionCommentThrowTag.WrongNumber -- this method _does_ throw a RouteException
+	// in convert_notices_to_exceptions.
 	/**
 	 * Validate all items in the cart and check for errors.
 	 *
-	 * @throws RouteException Exception if invalid data is detected.
+	 * @throws StockAvailabilityException Exception if invalid data is detected due to insufficient stock levels.
+	 * @throws RouteException Exception if invalid data is detected but not due to insufficient stock levels.
 	 */
 	public function validate_cart_items() {
 		$cart       = $this->get_cart_instance();
 		$cart_items = $this->get_cart_items();
 
+		$out_of_stock_products         = [];
+		$too_many_in_cart_products     = [];
+		$partial_out_of_stock_products = [];
+		$not_purchasable_products      = [];
+
 		foreach ( $cart_items as $cart_item_key => $cart_item ) {
-			$this->validate_cart_item( $cart_item );
+			try {
+				$this->validate_cart_item( $cart_item );
+			} catch ( TooManyInCartException $error ) {
+				$too_many_in_cart_products[] = $error;
+			} catch ( NotPurchasableException $error ) {
+				$not_purchasable_products[] = $error;
+			} catch ( PartialOutOfStockException $error ) {
+				$partial_out_of_stock_products[] = $error;
+			} catch ( OutOfStockException $error ) {
+				$out_of_stock_products[] = $error;
+			}
+		}
+
+		if (
```

Comment:
I think this section can be streamlined. If you create the WP_Error object regardless you can evaluate each error type separately, each using `$error->add` as needed, and then finally you would only need to do a check like this:

```
if ( $error->has_errors() ) {
			WC()->cart->calculate_totals();
			throw new StockAvailabilityException(
				'woocommerce_stock_availability_error',
				$error
			);
}
```